src/Microsoft.AspNet.Mvc.ViewFeatures/ModelStateDictionaryExtensions.cs

```diff
@@ -0,0 +1,115 @@
+        {
+            modelState.AddModelError(GetExpressionText(expression), errorMessage);
+        }
+
+        /// <summary>
+        /// Adds the specified <paramref name="exception"/> to the <see cref="ModelState.Errors"/> instance
+        /// that is associated with the specified <paramref name="expression"/>.
+        /// </summary>
+        /// <typeparam name="TModel">The type of the model.</typeparam>
+        /// <param name="modelState">The <see cref="ModelStateDictionary"/> instance this method extends.</param>
+        /// <param name="expression">An expression to be evaluated against an item in the current model.</param>
+        /// <param name="exception">The <see cref="Exception"/> to add.</param>
+        public static void AddModelError<TModel>(this ModelStateDictionary modelState, Expression<Func<TModel, object>> expression, Exception exception)
+        {
+            modelState.AddModelError(GetExpressionText(expression), exception);
+        }
+
+        /// <summary>
+        /// Removes the specified <paramref name="expression"/> from the <see cref="ModelStateDictionary"/>.
+        /// </summary>
+        /// <typeparam name="TModel">The type of the model.</typeparam>
+        /// <param name="modelState">The <see cref="ModelStateDictionary"/> instance this method extends.</param>
+        /// <param name="expression">An expression to be evaluated against an item in the current model.</param>
+        /// <returns>
+        /// true if the element is successfully removed; otherwise, false.
+        /// This method also returns false if <paramref name="expression"/> was not found in the model-state dictionary.
+        /// </returns>
+        public static bool Remove<TModel>(this ModelStateDictionary modelState, Expression<Func<TModel, object>> expression)
+        {
+            return modelState.Remove(GetExpressionText(expression));
+        }
+
+        /// <summary>
+        /// Removes all the entries for the specified <paramref name="expression"/> from the <see cref="ModelStateDictionary"/>.
+        /// </summary>
+        /// <typeparam name="TModel">The type of the model.</typeparam>
+        /// <param name="modelState">The <see cref="ModelStateDictionary"/> instance this method extends.</param>
+        /// <param name="expression">An expression to be evaluated against an item in the current model.</param>
+        public static void RemoveAll<TModel>(this ModelStateDictionary modelState, Expression<Func<TModel, object>> expression)
+        {
+            string modelKey = GetExpressionText(expression);
+            if (string.IsNullOrEmpty(modelKey))
+            {
+                var modelMetadata = new EmptyModelMetadataProvider().GetMetadataForType(typeof(TModel));
+                var elementMetadata = modelMetadata.ElementMetadata;
+                if (elementMetadata != null)
+                {
+                    modelMetadata = elementMetadata;
+                }
```

Comment:
I can't think of a case where property names of an element will appear in the `ModelStateDictionary` at the top level. A top-level collection of simple elements e.g. `string[]` might be bound as `string.Empty` (though HTML generation disallows the empty key) but will normally appear with prefixes such as `"[0]"`.

Suggest removing this block for now. Worst case is the block below will attempt to clean out entries with the `"Count"`, `"IsReadOnly"` or similar prefix. File a follow-up issue to find all collection prefixes and do the right thing. For example, we could eventually find the collection prefixes borrowing code from `JQueryFormValueProviderFactory.NormalizeJQueryToMvc()`. No incredible rush -- top-level collections are unusual.